packages/config-loader/src/lib/transform/apply.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { JsonObject, JsonValue } from '@backstage/config';
+import { TransformFunc } from './types';
+import { isObject } from './utils';
+
+/**
+ * Reads and parses, and validates, and transforms a single config file.
+ * The transformation rewrites any special values, like the $secret key.
+ */
+export async function applyConfigTransforms(
+  initialDir: string,
+  input: JsonValue,
+  transforms: TransformFunc[],
+): Promise<JsonObject> {
+  async function transform(
+    inputObj: JsonValue,
+    path: string,
+    baseDir: string,
+  ): Promise<JsonValue | undefined> {
+    let obj = inputObj;
+    let dir = baseDir;
+
+    for (const tf of transforms) {
+      try {
+        const result = await tf(inputObj, baseDir);
+        if (result.applied) {
+          if (result.value === undefined) {
+            return undefined;
+          }
+          obj = result.value;
+          dir = result.newBaseDir ?? dir;
+          break;
```

Comment:
Yeah not too keen on allowing substitutions of included files, there are some potential security issues there if used in a bad way, and not too useful either. Really the only thing that you'd want to do that doesn't work without re-applying the transforms is substitutions in included config. I don't think we want to support that for text files, but for included config files it could be a bit more expected to work. It's pretty easy to work around though and a bit of a weird usage, so I'm thinking we leave it like this and if it ends up being a problem we go for some kind of workaround.